sdks/go/pkg/beam/core/runtime/harness/statemgr.go

```diff
@@ -0,0 +1,310 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package harness
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/apache/beam/sdks/go/pkg/beam/core/runtime/exec"
+	"github.com/apache/beam/sdks/go/pkg/beam/log"
+	pb "github.com/apache/beam/sdks/go/pkg/beam/model/fnexecution_v1"
+	"github.com/golang/protobuf/proto"
+	"github.com/pkg/errors"
+)
+
+// ScopedStateManager scopes the global gRPC state manager to a single instruction.
+// The indirection makes it easier to control access.
+type ScopedSideInputReader struct {
+	mgr    *StateChannelManager
+	instID string
+
+	opened []io.Closer // track open readers to force close all
+	closed bool
+	mu     sync.Mutex
+}
+
+func NewScopedSideInputReader(mgr *StateChannelManager, instID string) *ScopedSideInputReader {
```

Comment:
Today no. But I think we should tease out the fnapi-facing code from the harness, so that alternative harnesses and runners could be built. That's why I'm leaning towards keeping this public similarly to a lot of the Exec stuff.